community/kernel/src/main/java/org/neo4j/kernel/recovery/LatestCheckPointFinder.java

```diff
@@ -91,9 +93,28 @@ public LatestCheckPoint find( long fromVersionBackwards ) throws IOException
 
             if ( latestCheckPoint != null )
             {
+                // Is the latest start entry in this log file version later than what the latest check point targets?
+                LogPosition target = latestCheckPoint.getLogPosition();
                 boolean commitsAfterCheckPoint = latestStartEntry != null &&
-                        latestStartEntry.getStartPosition().compareTo( latestCheckPoint.getLogPosition() ) >= 0;
-                return new LatestCheckPoint( latestCheckPoint, commitsAfterCheckPoint , oldestVersionFound );
+                        latestStartEntry.getStartPosition().compareTo( target ) >= 0;
+                if ( !commitsAfterCheckPoint )
+                {
+                    if ( target.getLogVersion() < version )
+                    {
+                        // This check point entry targets a previous log file.
+                        // Go there and see if there's a transaction. Reader is capped to that log version.
+                        commitsAfterCheckPoint = extractFirstTxIdAfterPosition( target, target.getLogVersion() ) !=
+                                LatestCheckPoint.NO_TRANSACTION_ID;
+                    }
+                }
+
+                // Extract first transaction id after check point target position.
+                // Reader may continue into log files after the initial version.
+                long firstTxIdAfterCheckPoint = commitsAfterCheckPoint
+                        ? extractFirstTxIdAfterPosition( target, fromVersionBackwards )
+                        : LatestCheckPoint.NO_TRANSACTION_ID;
+                return new LatestCheckPoint( latestCheckPoint, commitsAfterCheckPoint,
+                        firstTxIdAfterCheckPoint, oldestVersionFound );
```

Comment:
That block has two parts:

- decide whether there are commits after the last checkpoint
- which is the first transaction after the last checkpoint

my version tries to keep those apart. Your suggestions blends them together a bit, doesn't it? I too had some problems with it and found to separate them to be easier to reason about.